cdap-master/src/main/java/co/cask/cdap/data/tools/QueueConfigUpgrader.java

```diff
@@ -0,0 +1,133 @@
+import org.apache.hadoop.hbase.client.Put;
+import org.apache.hadoop.hbase.client.Result;
+import org.apache.hadoop.hbase.client.ResultScanner;
+import org.apache.hadoop.hbase.client.Scan;
+import org.apache.twill.filesystem.LocationFactory;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.List;
+import java.util.NavigableMap;
+import javax.annotation.Nullable;
+
+/**
+ * Upgrades queue.config table
+ */
+public class QueueConfigUpgrader extends AbstractUpgrader {
+  private static final Logger LOG = LoggerFactory.getLogger(QueueConfigUpgrader.class);
+  private final HBaseTableUtil tableUtil;
+
+  @Inject
+  public QueueConfigUpgrader(LocationFactory locationFactory, HBaseTableUtil tableUtil) {
+    super(locationFactory);
+    this.tableUtil = tableUtil;
+  }
+
+  @Override
+  void upgrade() throws Exception {
+    String dsName = Joiner.on(".").join(Constants.SYSTEM_NAMESPACE, QueueConstants.QUEUE_CONFIG_TABLE_NAME);
+    Id.DatasetInstance datasetId = Id.DatasetInstance.from(Constants.DEFAULT_NAMESPACE_ID, dsName);
+    HTable queueConfigTable = tableUtil.createHTable(HBaseConfiguration.create(),
+                                                     TableId.from(datasetId.getNamespaceId(), dsName));
+    LOG.info("Starting upgrade for queue config table {}", Bytes.toString(queueConfigTable.getTableName()));
+    try {
+      Scan scan = new Scan();
+      scan.setTimeRange(0, HConstants.LATEST_TIMESTAMP);
+      scan.addFamily(QueueEntryRow.COLUMN_FAMILY);
+      scan.setMaxVersions(1); // we only need to see one version of each row
+      ResultScanner resultScanner = queueConfigTable.getScanner(scan);
+      List<Put> puts = Lists.newArrayList();
+      List<Delete> deletes = Lists.newArrayList();
+      Result result;
+      try {
+        while ((result = resultScanner.next()) != null) {
+          byte[] row = result.getRow();
+          String rowKey = Bytes.toString(row);
+          LOG.info("Processing queue config for  {}", rowKey);
+          NavigableMap<byte[], byte[]> columnsMap = result.getFamilyMap(QueueEntryRow.COLUMN_FAMILY);
+          QueueName queueName = fromRowKey(row);
+          if (queueName != null) {
+            Put put = new Put(queueName.toBytes());
+            LOG.info("Upgraded queue name: {}", queueName);
+            for (NavigableMap.Entry<byte[], byte[]> entry : columnsMap.entrySet()) {
+              LOG.info("Adding entry {} -> {} for upgrade",
+                       Bytes.toString(entry.getKey()), Bytes.toString(entry.getValue()));
+              put.add(QueueEntryRow.COLUMN_FAMILY, entry.getKey(), entry.getValue());
+              puts.add(put);
+            }
+            LOG.info("Marking old key {} for deletion", rowKey);
+            deletes.add(new Delete(row));
+          }
+          LOG.info("Finished processing queue config for {}", rowKey);
+        }
+      } finally {
+        resultScanner.close();
+      }
+
+      queueConfigTable.put(puts);
```

Comment:
It's better to use `HTable.batch(List)`. You can represents both put and delete as `Mutation`.